File: biometric/src/main/java/dev/skomlach/biometric/compat/engine/internal/face/facelock/FacelockOldModule.java
package dev.skomlach.biometric.compat.engine.internal.face.facelock;

import android.app.admin.DevicePolicyManager;
import android.content.Context;
import android.content.pm.PackageManager;
import android.view.View;

import androidx.annotation.RestrictTo;
import androidx.core.os.CancellationSignal;

import java.lang.ref.WeakReference;

import dev.skomlach.biometric.compat.engine.AuthenticationFailureReason;
import dev.skomlach.biometric.compat.engine.BiometricInitListener;
import dev.skomlach.biometric.compat.engine.BiometricMethod;
import dev.skomlach.biometric.compat.engine.internal.AbstractBiometricModule;
import dev.skomlach.biometric.compat.engine.internal.core.Core;
import dev.skomlach.biometric.compat.engine.internal.core.interfaces.AuthenticationListener;
import dev.skomlach.biometric.compat.engine.internal.core.interfaces.RestartPredicate;
import dev.skomlach.biometric.compat.utils.BiometricErrorLockoutPermanentFix;
import dev.skomlach.biometric.compat.utils.LockType;
import dev.skomlach.biometric.compat.utils.logging.BiometricLoggerImpl;

@RestrictTo(RestrictTo.Scope.LIBRARY)
public class FacelockOldModule extends AbstractBiometricModule {

    public static final int BIOMETRIC_AUTHENTICATION_FAILED = 1001;
    private FaceLockHelper faceLockHelper = null;
    private ProxyListener facelockProxyListener = null;
    private WeakReference<View> viewWeakReference = new WeakReference<>(null);
    private BiometricInitListener listener;
    private boolean isConnected = false;

    public FacelockOldModule(BiometricInitListener initlistener) {
        super(BiometricMethod.FACELOCK);
        this.listener = initlistener;

        final FaceLockInterface faceLockInterface = new FaceLockInterface() {
            @Override
            public void onError(int code, String msg) {
                BiometricLoggerImpl.d("FaceIdModule" + ("FaceIdInterface.onError " + code + " " + msg));
                if (facelockProxyListener != null) {
                    int failureReason = BIOMETRIC_ERROR_CANCELED;
                    switch (code) {
                        case FaceLockHelper.FACELOCK_FAILED_ATTEMPT:
                            failureReason = BIOMETRIC_AUTHENTICATION_FAILED;

                            break;
                        case FaceLockHelper.FACELOCK_TIMEOUT:
                            failureReason = BIOMETRIC_ERROR_TIMEOUT;

                            break;
                        case FaceLockHelper.FACELOCK_NO_FACE_FOUND:
                            failureReason = BIOMETRIC_ERROR_UNABLE_TO_PROCESS;
                            break;
                        case FaceLockHelper.FACELOCK_NOT_SETUP:
                            failureReason = BIOMETRIC_ERROR_NO_BIOMETRICS;
                            break;
                        case FaceLockHelper.FACELOCK_CANCELED:
                            failureReason = BIOMETRIC_ERROR_CANCELED;
                            break;
                        case FaceLockHelper.FACELOCK_CANNT_START:
                        case FaceLockHelper.FACELOCK_UNABLE_TO_BIND:
                        case FaceLockHelper.FACELOCK_API_NOT_FOUND:
                            failureReason = BIOMETRIC_ERROR_HW_UNAVAILABLE;
                            break;
                    }
                    facelockProxyListener.onAuthenticationError(failureReason, msg);
                }
            }

            @Override
            public void onAuthorized() {
                BiometricLoggerImpl.d("FaceIdModule" + "FaceIdInterface.onAuthorized");
                if (facelockProxyListener != null) {
                    facelockProxyListener.onAuthenticationSucceeded(null);
                }
            }

            @Override
            public void onConnected() {
                BiometricLoggerImpl.d("FaceIdModule" + "FaceIdInterface.onConnected");
                if (facelockProxyListener != null) {
                    facelockProxyListener.onAuthenticationAcquired(0);
                }

                if (listener != null) {
                    isConnected = true;
                    listener.initFinished(BiometricMethod.FACELOCK, FacelockOldModule.this);
                    listener = null;
                    faceLockHelper.stopFaceLock();
                } else {
                    BiometricLoggerImpl.d("FaceIdModule" + ("authorize: " + viewWeakReference.get()));
                    faceLockHelper.startFaceLockWithUi(viewWeakReference.get());
                }
            }

            @Override
            public void onDisconnected() {
                BiometricLoggerImpl.d("FaceIdModule" + "FaceIdInterface.onDisconnected");
                if (facelockProxyListener != null) {
                    facelockProxyListener.onAuthenticationError(BIOMETRIC_ERROR_CANCELED,
                            FaceLockHelper.getMessage(BIOMETRIC_ERROR_CANCELED));
                }
                if (listener != null) {
                    listener.initFinished(BiometricMethod.FACELOCK, FacelockOldModule.this);
                    listener = null;
                    faceLockHelper.stopFaceLock();
                }
            }
        };

        faceLockHelper = new FaceLockHelper(getContext(), faceLockInterface);
        if (!isHardwarePresent()) {
            if (listener != null) {
                listener.initFinished(BiometricMethod.FACELOCK, FacelockOldModule.this);
                listener = null;
            }
            return;
        } else {
            faceLockHelper.initFacelock();
        }
    }

    public void stopAuth() {
        faceLockHelper.stopFaceLock();
        faceLockHelper.destroy();
    }

    @Override
    public boolean isManagerAccessible() {
        return isConnected;
    }

    @Override
    public boolean isHardwarePresent() {
        // Retrieve all services that can match the given intent
        if (!faceLockHelper.faceUnlockAvailable())
            return false;

        final PackageManager pm = getContext().getPackageManager();
        if (!pm.hasSystemFeature(PackageManager.FEATURE_CAMERA_FRONT)) {
            return false;
        }

        DevicePolicyManager dpm = (DevicePolicyManager) getContext().getSystemService(Context.DEVICE_POLICY_SERVICE);
        if (dpm.getCameraDisabled(null)) {
            return false;
        }

        return hasEnrolled();
    }

    @Override
    public boolean hasEnrolled() throws SecurityException {
        return LockType.isBiometricWeakEnabled(getContext());
    }

    @Override
    public void authenticate(final CancellationSignal cancellationSignal,
                             final AuthenticationListener listener,
                             final RestartPredicate restartPredicate) throws SecurityException {

        try {
            BiometricLoggerImpl.d("FaceIdModule" + "Facelock call authorize");
            authorize(new ProxyListener(restartPredicate, cancellationSignal, listener));
            return;
        } catch (Throwable e) {
            BiometricLoggerImpl.e(e, "BiometricGenericModule: authenticate failed unexpectedly");
        }

        if (listener != null) {
            listener.onFailure(AuthenticationFailureReason.UNKNOWN,
                    tag());
        }
    }

    public void setCallerView(View targetView) {
        BiometricLoggerImpl.d("FaceIdModule" + ("setCallerView: " + targetView));
        viewWeakReference = new WeakReference<>(targetView);
    }

    private void authorize(ProxyListener proxyListener) {
        this.facelockProxyListener = proxyListener;
        faceLockHelper.initFacelock();
    }

    @RestrictTo(RestrictTo.Scope.LIBRARY)
    public class ProxyListener {

        private final RestartPredicate restartPredicate;
        private final CancellationSignal cancellationSignal;
        private final AuthenticationListener listener;

        public ProxyListener(RestartPredicate restartPredicate,
                             CancellationSignal cancellationSignal, AuthenticationListener listener) {
            this.restartPredicate = restartPredicate;
            this.cancellationSignal = cancellationSignal;
            this.listener = listener;
        }

        public Void onAuthenticationError(int errMsgId, CharSequence errString) {
            AuthenticationFailureReason failureReason = AuthenticationFailureReason.UNKNOWN;
            switch (errMsgId) {
                case BIOMETRIC_ERROR_NO_BIOMETRICS:
                    failureReason = AuthenticationFailureReason.NO_BIOMETRICS_REGISTERED;
                    break;
                case BIOMETRIC_AUTHENTICATION_FAILED:
                    failureReason = AuthenticationFailureReason.AUTHENTICATION_FAILED;
                    break;
                case BIOMETRIC_ERROR_HW_NOT_PRESENT:
                    failureReason = AuthenticationFailureReason.NO_HARDWARE;
                    break;
                case BIOMETRIC_ERROR_HW_UNAVAILABLE:
                    failureReason = AuthenticationFailureReason.HARDWARE_UNAVAILABLE;
                    break;
                case BIOMETRIC_ERROR_LOCKOUT_PERMANENT:
                    BiometricErrorLockoutPermanentFix.INSTANCE.setBiometricSensorPermanentlyLocked(getType());
                    failureReason = AuthenticationFailureReason.HARDWARE_UNAVAILABLE;
                    break;
                case BIOMETRIC_ERROR_UNABLE_TO_PROCESS:
                case BIOMETRIC_ERROR_NO_SPACE:
                    failureReason = AuthenticationFailureReason.SENSOR_FAILED;
                    break;
                case BIOMETRIC_ERROR_TIMEOUT:
                    failureReason = AuthenticationFailureReason.TIMEOUT;
                    break;
                case BIOMETRIC_ERROR_LOCKOUT:
                    lockout();
                    failureReason = AuthenticationFailureReason.LOCKED_OUT;
                    break;
                case BIOMETRIC_ERROR_USER_CANCELED:
                    Core.cancelAuthentication(FacelockOldModule.this);
                    return null;
                case BIOMETRIC_ERROR_CANCELED:
                    // Don't send a cancelled message.
                    return null;
            }

            if (restartPredicate.invoke(failureReason)) {
                if (listener != null) {
                    listener.onFailure(failureReason, tag());
                }
                authenticate(cancellationSignal, listener, restartPredicate);
            } else {
                switch (failureReason) {
                    case SENSOR_FAILED:
                    case AUTHENTICATION_FAILED:
                        lockout();
                        failureReason = AuthenticationFailureReason.LOCKED_OUT;
                        break;
                }
                if (listener != null) {
                    listener.onFailure(failureReason, tag());
                }
            }
            return null;
        }

        public Void onAuthenticationHelp(int helpMsgId, CharSequence helpString) {
            return null;
        }

        public Void onAuthenticationSucceeded(Object result) {
            if (listener != null) {
                listener.onSuccess(tag());
            }
            return null;
        }

        public Void onAuthenticationAcquired(int acquireInfo) {
            BiometricLoggerImpl.d("FaceIdModule" + ("FaceIdInterface.ProxyListener " + acquireInfo));
            return null;
        }

        public Void onAuthenticationFailed() {
            if (listener != null) {
                listener.onFailure(AuthenticationFailureReason.AUTHENTICATION_FAILED,
                        tag());
            }
            return null;
        }
    }
}

File: biometric/src/main/java/dev/skomlach/biometric/compat/engine/internal/face/facelock/FacelockOldModule.kt
/*
 *  Copyright (c) 2023 Sergey Komlach aka Salat-Cx65; Original project https://github.com/Salat-Cx65/AdvancedBiometricPromptCompat
 *  All rights reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

package dev.skomlach.biometric.compat.engine.internal.face.facelock

import android.app.admin.DevicePolicyManager
import android.content.Context
import android.content.pm.PackageManager
import android.view.SurfaceHolder
import android.view.SurfaceView
import android.view.View
import androidx.core.os.CancellationSignal
import dev.skomlach.biometric.compat.AuthenticationFailureReason
import dev.skomlach.biometric.compat.BiometricCryptoObject
import dev.skomlach.biometric.compat.engine.BiometricInitListener
import dev.skomlach.biometric.compat.engine.BiometricMethod
import dev.skomlach.biometric.compat.engine.core.Core
import dev.skomlach.biometric.compat.engine.core.interfaces.AuthenticationListener
import dev.skomlach.biometric.compat.engine.core.interfaces.RestartPredicate
import dev.skomlach.biometric.compat.engine.internal.AbstractBiometricModule
import dev.skomlach.biometric.compat.utils.LockType.isBiometricWeakEnabled
import dev.skomlach.biometric.compat.utils.logging.BiometricLoggerImpl.d
import dev.skomlach.biometric.compat.utils.logging.BiometricLoggerImpl.e
import dev.skomlach.common.misc.ExecutorHelper
import java.lang.ref.WeakReference


class FacelockOldModule(private var listener: BiometricInitListener?) :
    AbstractBiometricModule(BiometricMethod.FACELOCK) {
    private var faceLockHelper: FaceLockHelper? = null
    private var facelockProxyListener: ProxyListener? = null
    private var viewWeakReference = WeakReference<SurfaceView?>(null)
    override var isManagerAccessible = false


    init {
        val faceLockInterface: FaceLockInterface = object : FaceLockInterface {
            override fun onError(code: Int, msg: String) {
                d("$name:FaceIdInterface.onError $code $msg")
                if (facelockProxyListener != null) {
                    facelockProxyListener?.onAuthenticationError(code, msg)
                }
            }

            override fun onAuthorized() {
                d("$name.FaceIdInterface.onAuthorized")
                if (facelockProxyListener != null) {
                    facelockProxyListener?.onAuthenticationSucceeded(null)
                }
            }

            override fun onConnected() {
                d("$name.FaceIdInterface.onConnected")
                if (facelockProxyListener != null) {
                    facelockProxyListener?.onAuthenticationAcquired(0)
                }
                if (listener != null) {
                    isManagerAccessible = true
                    listener?.initFinished(biometricMethod, this@FacelockOldModule)
                    listener = null
                    faceLockHelper?.stopFaceLock()
                } else {
                    try {
                        d(name + ".authorize: " + viewWeakReference.get())
                        viewWeakReference.get()?.let { view ->
                            if (view.visibility == View.VISIBLE || view.holder.isCreating) {
                                authCallTimestamp.set(System.currentTimeMillis())
                                faceLockHelper?.startFaceLockWithUi(view)
                                return
                            } else {
                                view.holder.addCallback(object : SurfaceHolder.Callback {
                                    override fun surfaceCreated(p0: SurfaceHolder) {
                                        authCallTimestamp.set(System.currentTimeMillis())
                                        faceLockHelper?.startFaceLockWithUi(view)
                                    }

                                    override fun surfaceChanged(
                                        p0: SurfaceHolder,
                                        p1: Int,
                                        p2: Int,
                                        p3: Int
                                    ) {

                                    }

                                    override fun surfaceDestroyed(p0: SurfaceHolder) {

                                    }
                                })
                                view.visibility = View.VISIBLE
                            }

                        } ?: kotlin.run {
                            authCallTimestamp.set(System.currentTimeMillis())
                            faceLockHelper?.startFaceLockWithUi(null)
                        }

                    } catch (e: Throwable) {
                        e("$name.FaceIdInterface.onConnected", e)
                    }
                }
            }

            override fun onDisconnected() {
                d("$name.FaceIdInterface.onDisconnected")
                if (facelockProxyListener != null) {
                    facelockProxyListener?.onAuthenticationError(
                        FaceLockHelper.FACELOCK_CANCELED,
                        FaceLockHelper.getMessage(FaceLockHelper.FACELOCK_CANCELED)
                    )
                }
                if (listener != null) {
                    listener?.initFinished(biometricMethod, this@FacelockOldModule)
                    listener = null
                    faceLockHelper?.stopFaceLock()
                }
            }
        }
        faceLockHelper = FaceLockHelper(faceLockInterface)
        if (!isHardwarePresent) {
            if (listener != null) {
                listener?.initFinished(biometricMethod, this@FacelockOldModule)
                listener = null
            }
        } else {
            faceLockHelper?.initFacelock()
        }
    }

    override val isUserAuthCanByUsedWithCrypto: Boolean
        get() = false

    fun stopAuth() {
        faceLockHelper?.stopFaceLock()
        faceLockHelper?.destroy()
    }

    override fun getManagers(): Set<Any> {
        //No way to detect enrollments
        return emptySet()
    }

    // Retrieve all services that can match the given intent
    override val isHardwarePresent: Boolean
        get() {
            val pm = context.packageManager
            if (!pm.hasSystemFeature(PackageManager.FEATURE_CAMERA_FRONT)) {
                return false
            }
            val dpm =
                context.getSystemService(Context.DEVICE_POLICY_SERVICE) as DevicePolicyManager?
            if (dpm?.getCameraDisabled(null) == true)
                return false

            // Retrieve all services that can match the given intent
            return faceLockHelper?.faceUnlockAvailable() == true
        }

    override val hasEnrolled: Boolean
        get() {
            return isBiometricWeakEnabled("com.android.facelock", context)
        }

    @Throws(SecurityException::class)
    override fun authenticate(
        biometricCryptoObject: BiometricCryptoObject?,
        cancellationSignal: CancellationSignal?,
        listener: AuthenticationListener?,
        restartPredicate: RestartPredicate?
    ) {
        d("$name.authenticate - $biometricMethod; Crypto=$biometricCryptoObject")
        try {
            d("$name: Facelock call authorize")
            cancellationSignal?.setOnCancelListener {
                stopAuth()
            }
            authorize(
                ProxyListener(
                    biometricCryptoObject,
                    restartPredicate,
                    cancellationSignal,
                    listener
                )
            )
            return
        } catch (e: Throwable) {
            e(e, "$name: authenticate failed unexpectedly")
        }
        listener?.onFailure(
            AuthenticationFailureReason.UNKNOWN,
            tag()
        )
    }

    fun setCallerView(targetView: SurfaceView?) {
        d("$name.setCallerView: $targetView")
        viewWeakReference = WeakReference(targetView)
    }

    private fun authorize(proxyListener: ProxyListener) {
        facelockProxyListener = proxyListener
        faceLockHelper?.stopFaceLock()
        faceLockHelper?.initFacelock()
    }


    inner class ProxyListener(
        private val biometricCryptoObject: BiometricCryptoObject?,
        private val restartPredicate: RestartPredicate?,
        private val cancellationSignal: CancellationSignal?,
        private val listener: AuthenticationListener?
    ) {
        private var errorTs = System.currentTimeMillis()
        private val skipTimeout =
            context.resources.getInteger(android.R.integer.config_shortAnimTime)
        private var selfCanceled = false
        fun onAuthenticationError(errMsgId: Int, errString: CharSequence?): Void? {
            d("$name.onAuthenticationError: $errMsgId-$errString")
            val tmp = System.currentTimeMillis()
            if (tmp - errorTs <= skipTimeout || tmp - authCallTimestamp.get() <= skipTimeout)
                return null
            errorTs = tmp
            var failureReason = AuthenticationFailureReason.UNKNOWN
            when (errMsgId) {
                FaceLockHelper.FACELOCK_FAILED_ATTEMPT -> failureReason =
                    AuthenticationFailureReason.AUTHENTICATION_FAILED

                FaceLockHelper.FACELOCK_TIMEOUT -> failureReason =
                    AuthenticationFailureReason.TIMEOUT

                FaceLockHelper.FACELOCK_NO_FACE_FOUND -> failureReason =
                    AuthenticationFailureReason.HARDWARE_UNAVAILABLE

                FaceLockHelper.FACELOCK_NOT_SETUP -> failureReason =
                    AuthenticationFailureReason.NO_BIOMETRICS_REGISTERED

                FaceLockHelper.FACELOCK_CANNT_START, FaceLockHelper.FACELOCK_UNABLE_TO_BIND, FaceLockHelper.FACELOCK_API_NOT_FOUND -> failureReason =
                    AuthenticationFailureReason.HARDWARE_UNAVAILABLE

                else -> {
                    if (!selfCanceled) {
                        listener?.onFailure(failureReason, tag())
                        postCancelTask {
                            if (cancellationSignal?.isCanceled == false) {
                                selfCanceled = true
                                listener?.onCanceled(tag())
                                Core.cancelAuthentication(this@FacelockOldModule)
                            }
                        }
                    }
                    return null
                }
            }
            if (restartCauseTimeout(failureReason)) {
                selfCanceled = true
                stopAuth()
                ExecutorHelper.postDelayed({
                    authenticate(
                        biometricCryptoObject,
                        cancellationSignal,
                        listener,
                        restartPredicate
                    )
                }, skipTimeout.toLong())
            } else
                if (failureReason == AuthenticationFailureReason.TIMEOUT || restartPredicate?.invoke(
                        failureReason
                    ) == true
                ) {
                    listener?.onFailure(failureReason, tag())
                    selfCanceled = true
                    stopAuth()
                    ExecutorHelper.postDelayed({
                        authenticate(
                            biometricCryptoObject,
                            cancellationSignal,
                            listener,
                            restartPredicate
                        )
                    }, 1000)
                } else {
                    if (mutableListOf(
                            AuthenticationFailureReason.SENSOR_FAILED,
                            AuthenticationFailureReason.AUTHENTICATION_FAILED
                        ).contains(failureReason)
                    ) {
                        lockout()
                        failureReason = AuthenticationFailureReason.LOCKED_OUT
                    }
                    listener?.onFailure(failureReason, tag())
                    postCancelTask {
                        if (cancellationSignal?.isCanceled == false) {
                            selfCanceled = true
                            listener?.onCanceled(tag())
                            Core.cancelAuthentication(this@FacelockOldModule)
                        }
                    }
                }
            return null
        }

        fun onAuthenticationHelp(helpMsgId: Int, helpString: CharSequence?): Void? {
            d("$name.onAuthenticationHelp: $helpMsgId-$helpString")
            return null
        }

        fun onAuthenticationSucceeded(result: Any?): Void? {
            d("$name.onAuthenticationSucceeded $result")
            val tmp = System.currentTimeMillis()
            if (tmp - errorTs <= skipTimeout || tmp - authCallTimestamp.get() <= skipTimeout)
                return null
            errorTs = tmp
            listener?.onSuccess(
                tag(),
                BiometricCryptoObject(
                    biometricCryptoObject?.signature,
                    biometricCryptoObject?.cipher,
                    biometricCryptoObject?.mac
                )
            )
            return null
        }

        fun onAuthenticationAcquired(acquireInfo: Int): Void? {
            d("$name.onAuthenticationAcquired $acquireInfo")
            return null
        }

        fun onAuthenticationFailed(): Void? {
            d("$name.onAuthenticationFailed")
            val tmp = System.currentTimeMillis()
            if (tmp - errorTs <= skipTimeout || tmp - authCallTimestamp.get() <= skipTimeout)
                return null
            errorTs = tmp
            listener?.onFailure(
                AuthenticationFailureReason.AUTHENTICATION_FAILED,
                tag()
            )
            return null
        }
    }
}